keeper/components/dispatcher_state_test.go

```diff
@@ -0,0 +1,440 @@
+
+	hasDispatched := dispatcher.HasDispatchedAmount(ctx, sourceInfo, destinationInfo, denom)
+	require.False(t, hasDispatched)
+
+	totalDispatched := dispatcher.GetDispatchedAmountsByProtocolID(ctx, types.PROTOCOL_IBC)
+	require.NotNil(t, totalDispatched.ChainAmount)
+	require.Empty(t, totalDispatched.ChainAmount)
+
+	// Test iteration with empty state
+	called := false
+	dispatcher.IterateDispatchedAmountsByProtocolID(
+		ctx,
+		types.PROTOCOL_IBC,
+		func(sourceCounterpartyId string, dispatchedInfo types.ChainAmountDispatched) bool {
+			called = true
+			return false
+		},
+	)
+	require.False(t, called)
+}
+
+func TestDispatched_MultipleProtocolsAndChains(t *testing.T) {
+	dispatcher, deps := newDispatcherKeeper(t)
+	ctx := deps.SdkCtx
+
+	// Test with multiple protocols and chains
+	testCases := []struct {
+		sourceInfo      types.OrbitID
+		destinationInfo types.OrbitID
+		denom           string
+		amount          types.AmountDispatched
+	}{
+		{
+			sourceInfo: types.OrbitID{
+				ProtocolID:     types.PROTOCOL_IBC,
+				CounterpartyID: "channel-1",
+			},
+			destinationInfo: types.OrbitID{
+				ProtocolID:     types.PROTOCOL_CCTP,
+				CounterpartyID: "0",
+			},
+			denom: "uusdc",
+			amount: types.AmountDispatched{
+				Incoming: math.NewInt(100),
+				Outgoing: math.NewInt(50),
+			},
+		},
+		{
+			sourceInfo: types.OrbitID{
+				ProtocolID:     types.PROTOCOL_IBC,
+				CounterpartyID: "channel-2",
+			},
+			destinationInfo: types.OrbitID{
+				ProtocolID:     types.PROTOCOL_CCTP,
+				CounterpartyID: "1",
+			},
+			denom: "uusdc",
+			amount: types.AmountDispatched{
+				Incoming: math.NewInt(200),
+				Outgoing: math.NewInt(100),
+			},
+		},
+		{
+			sourceInfo: types.OrbitID{
+				ProtocolID:     types.PROTOCOL_CCTP,
+				CounterpartyID: "0",
+			},
+			destinationInfo: types.OrbitID{
+				ProtocolID:     types.PROTOCOL_IBC,
+				CounterpartyID: "channel-3",
+			},
+			denom: "uusdc",
+			amount: types.AmountDispatched{
+				Incoming: math.NewInt(300),
+				Outgoing: math.NewInt(150),
+			},
+		},
+	}
+
+	// Set all dispatch records
+	for _, tc := range testCases {
+		err := dispatcher.SetDispatchedAmount(
+			ctx,
+			tc.sourceInfo,
+			tc.destinationInfo,
+			tc.denom,
+			tc.amount,
+		)
+		require.NoError(t, err)
+	}
+
+	// Verify all records can be retrieved
+	for _, tc := range testCases {
+		result := dispatcher.GetDispatchedAmount(ctx, tc.sourceInfo, tc.destinationInfo, tc.denom)
+		require.Equal(t, tc.amount, result)
+
+		hasDispatched := dispatcher.HasDispatchedAmount(
+			ctx,
+			tc.sourceInfo,
+			tc.destinationInfo,
+			tc.denom,
+		)
+		require.True(t, hasDispatched)
+	}
+
+	// Test protocol-specific queries
+	ibcTotal := dispatcher.GetDispatchedAmountsByProtocolID(ctx, types.PROTOCOL_IBC)
+	require.Len(t, ibcTotal.ChainAmount, 2) // channel-1 and channel-2
+
+	cctpTotal := dispatcher.GetDispatchedAmountsByProtocolID(ctx, types.PROTOCOL_CCTP)
+	require.Len(t, cctpTotal.ChainAmount, 1) // only counterparty "0"
```

Comment:
what's the distinction between `GetDispatchedAmountsByProtocolID` and `GetDispatchedAmountsByDestinationProtocolID`? If two chain are connected via a given technology then that'll be the same `ProtocolID` for both source and destination no?